lib/utils/concurrentqueue/queue.go

```diff
@@ -0,0 +1,280 @@
+}
+
+// run spawns background tasks and then blocks on collection/reordering routine.
+func (q *Queue) run(workfn func(interface{}) interface{}, cfg config) {
+	// internal worker input/output channels. due to the semaphore channel below,
+	// sends on these channels never block, as they are each allocated with sufficient
+	// capacity to hold all in-flight items.
+	workerIn, workerOut := make(chan item, cfg.capacity), make(chan item, cfg.capacity)
+
+	// semaphore channel used to limit the number of "in flight" items.  a message is added prior to accepting
+	// every input and removed upon emission of every output.  this allows us to exert backpressure and prevent
+	// uncapped memory growth due to a slow worker.  this also keeps the queue's "capacity" consistent, regardless
+	// of whether we are exeriencing general slowness or a "head of line blocking" type scenario.
+	sem := make(chan struct{}, cfg.capacity)
+
+	// spawn workers
+	for i := 0; i < cfg.workers; i++ {
+		go func() {
+			for {
+				var itm item
+				select {
+				case itm = <-workerIn:
+				case <-q.done:
+					return
+				}
+
+				itm.value = workfn(itm.value)
+
+				select {
+				case workerOut <- itm:
+				default:
+					panic("cq worker output channel already full (semaphore violation)")
+				}
+			}
+		}()
+	}
+
+	go q.distribute(workerIn, sem)
+
+	q.collect(workerOut, sem)
+}
+
+// distribute takes inbound work items, applies a nonce, and then distributes
+// them to the workers.
+func (q *Queue) distribute(workerIn chan<- item, sem chan struct{}) {
+	var nonce uint64
+	for {
+		// we are about to accept an input, add an item to the in-flight semaphore channel
+		select {
+		case sem <- struct{}{}:
+		case <-q.done:
+			return
+		}
+
+		var value interface{}
+		select {
+		case value = <-q.input:
+		case <-q.done:
+			return
+		}
+
+		select {
+		case workerIn <- item{value: value, nonce: nonce}:
+		default:
+			panic("cq worker input channel already full (semaphore violation)")
+		}
+
+		nonce++
+	}
+}
+
+// collect takes the potentially disordered worker output and unifies it into
+// an ordered output.
+func (q *Queue) collect(workerOut <-chan item, sem chan struct{}) {
+	// items that cannot be emitted yet (due to arriving out of order),
+	// stored in mapping of nonce => value.
+	queue := make(map[uint64]interface{})
+
+	// the nonce of the item we need to emit next.  incremented upon
+	// successful emission.
+	var nonce uint64
+
+	// output value to be emitted (if any).  note that nil is a valid
+	// output value, so we cannot inspect this value directly to
+	// determine our state.
+	var out interface{}
+
+	// emit indicates whether or not we should be attempting to emit
+	// the output value.
+	var emit bool
+
+	for {
+		outc := q.output
+		if !emit {
+			// we do not have the next output item yet, do not attempt to send
+			outc = nil
+		}
+
+		select {
+		case itm := <-workerOut:
+			if itm.nonce == nonce {
+				// item matches current nonce, send it immdeiately
```

Comment:
The way it's currently implemented, it might not be sent immediately unless I'm missing something? It will go to the next for loop iteration and if there are other items in `workerOut`, this select may just keep accumulating them in `queue` (IIRC select case order is not defined).

Reply:
You're right, it isn't sent immediately, the loop just goes immediately into a sending state rather than enqueuing the value.  I've updated the comments to be more clear.

IMO output channel starvation isn't a serious risk here.  Any non-trivial work function will leave this select statement idle (i.e. able to emit events) the majority of the time.  Any flow of events arriving so fast as to be a serious hindrance to event emission will hit the semaphore limit almost immediately, at which point emission will succeed anyhow.